doc/likwid-pin.1
.TH LIKWID-PIN 1 <DATE> likwid\-VERSION
.SH NAME
likwid-pin \- pin a sequential or threaded application to dedicated processors
.SH SYNOPSIS
.B likwid-pin
.RB [\-vhSpqim]
.RB [ \-V
.IR <verbosity> ]
.RB [ \-c/\-C
.IR <corelist> ]
.RB [ \-s
.IR <skip_mask> ]
.RB [ \-d
.IR <delim> ]
.SH DESCRIPTION
.B likwid-pin
is a command line application to pin a sequential or multithreaded
application to dedicated processors. It can be used as replacement for taskset.
Opposite to taskset no affinity mask but single processors are specified.
For multithreaded applications based on the pthread library the
.B pthread_create
library call is overloaded through LD_PRELOAD and each created thread is pinned
to a dedicated processor as specified in
.I core_list .
.PP
Per default every generated thread is pinned to the core in the order of calls to
.B pthread_create
it is possible to skip single threads.
.PP
The OpenMP implementations of GCC and ICC compilers are explicitly supported.
Clang's OpenMP backend should also work as it is built on top of Intel's OpenMP runtime library.
Others may also work
.B likwid-pin
sets the environment variable
.B OMP_NUM_THREADS
for you if not already present.
It will set as many threads as present in the pin expression. Be aware that
with pthreads the parent thread is always pinned. If you create for example 4
threads with
.B pthread_create
and do not use the parent process as worker you still have to provide
.B num_threads+1
processor ids.
.PP
.B likwid-pin
supports different numberings for pinning. See section
.B CPU EXPRESSION
for details.
.PP
For applications where first touch policy on NUMA systems cannot be employed
.B likwid-pin
can be used to turn on interleave memory placement. This can significantly
speed up the performance of memory bound multithreaded codes. All NUMA nodes
the user pinned threads to are used for interleaving.

LIKWID introduces the concept of affinity domains which can be described as all hardware threads sharing a topological
entity. There are, for example, affinity domains for each CPU socket in the system and each of these affinity domains contains
the hardware threads that belong to the socket. There is always a virtual affinity domain 'N' for the whole node, thus all hardware threads. Further names are 'Sx' with 'x' as socket offset ('S0', 'S1', 'S2', ...), 'Dy' with 'y' as CPU die offset, 'Mz' with 'z' as NUMA domain offset and 'Cx' with 'x' as last level cache offset. It is the offset of the topological entities, so when there are two sockets with IDs 74 und 10023 (systems like this exist), the socket affinity domains are 'S0' and 'S1'.

.SH OPTIONS
.TP
.B \-\^h,\-\-\^help
prints a help message to standard output, then exits.
.TP
.B \-\^v,\-\-\^version
prints version information to standard output, then exits.
.TP
.B \-\^V, \-\-\^verbose <level>
verbose output during execution for debugging. 0 for only errors, 1 for informational output, 2 for detailed output and 3 for developer output
.TP
.B \-\^c,\-\^C <cpu expression>
specify a numerical list of processors. The list may contain multiple  items, separated by comma, and ranges. For example 0,3,9-11. Other format are available, see the
.B CPU EXPRESSION
section.
.TP
.B \-\^s, \-\-\^skip <skip_mask>
Specify skip mask as HEX number. For each set bit the corresponding thread is skipped.
.TP
.B \-\^S,\-\-\^sweep
All ccNUMA memory domains belonging to the specified thread list will be cleaned before the run. Can solve file buffer cache problems on Linux.
.TP
.B \-\^p
prints the available thread domains for logical pinning
.TP
.B \-\^i
set NUMA memory policy to interleave involving all NUMA nodes involved in pinning
.TP
.B \-\^m
set NUMA memory policy to membind involving all NUMA nodes involved in pinning
.TP
.B \-\^d <delim>
usable with
.B \-\^p
to specify the CPU delimiter in the cpulist
.TP
.B \-\^q,\-\-\^quiet
silent execution without output

.SH CPU EXPRESSION
.IP 1. 4
The most intuitive CPU selection method is a comma-separated list of hardware thread IDs. An example for this is
.B 0,2
which schedules the threads on hardware threads 
.B 0
and
.B 2.
The physical numbering also allows the usage of ranges like
.B 0-2
which results in the list
.B 0,1,2.
.IP 2. 4
The CPUs can be selected by their indices inside of an affinity domain. The format is
.B L:<domain>:<indexlist>
for selecting the CPUs inside the given domain. Assuming an virtual affinity domain
.B 'P'
that contains the CPUs
.B 0,4,1,5,2,6,3,7.
After sorting it to have physical hardware threads first we get:
.B 0,1,2,3,4,5,6,7.
The logical numbering
.B L:P:0-2
results in the selection
.B 0,1,2
from the physical hardware threads first list.
.IP 3. 4
The expression syntax enables the selection according to an selection function with variable input parameters. The format is either
.B E:<affinity domain>:<numberOfThreads>
to use the first <numberOfThreads> threads in affinity domain <affinity domain> or
.B E:<affinity domain>:<numberOfThreads>:<chunksize>:<stride>
to use <numberOfThreads> threads with <chunksize> threads selected in row while skipping <stride> threads in affinity domain <affinity domain>. Examples are
.B E:N:4:1:2
for selecting the first four physical CPUs on a system with 2 hardware thread per CPU core or
.B E:P:4:2:4
for choosing the first two threads in affinity domain
.B P,
skipping 2 threads and selecting again two threads. The resulting CPU list for virtual affinity domain
.B P
is
.B 0,4,2,6
.IP 3. 4
The last format schedules the threads not only in a single affinity domain but distributed them evenly over all available affinity domains of the same kind. In contrast to the other formats, the selection is done using the physical hardware threads first and then the virtual hardware threads (aka SMT threads). The format is
.B <affinity domain without number>:scatter
like
.B M:scatter
to schedule the threads evenly in all available memory affinity domains. Assuming the two socket domains
.B S0 = 0,4,1,5
and
.B S1 = 2,6,3,7
the expression
.B S:scatter
results in the CPU list
.B 0,2,1,3,4,6,5,7
Besides scatter, there is also 'balanced' (close pinning over domains in natural order) and 'cbalanced' (close pinning over domain with physical cores first). If we assume the socket setup above, the expression
.B S:balanced:4
results in the CPU list
.B 0,4,2,6
while with cbalanced, the CPU list is (assuming 4-7 being SMT threads)
.B 0,1,2,3

.SH EXAMPLE
.IP 1. 5
For standard pthread application:
.TP
.B likwid-pin -c 0,2,4-6 ./myApp
.PP
The parent process is pinned to processor 0 which is likely to be thread 0 in
.B ./myApp.
Thread 1 is pinned to processor 2, thread 2 to processor 4, thread 3 to processor 5 and thread 4 to processor 6. If more threads
are created than specified in the processor list, these threads are pinned to processor 0 as fallback.
.IP 2. 5
For selection of CPUs inside of a CPUset only the logical numbering is allowed. Assuming CPUset
.B 0,4,1,5:
.TP
.B likwid-pin -c L:N:1,3 ./myApp
.PP
This command pins
.B ./myApp
on CPU
.B 4
and the thread started by
.B ./myApp
on CPU
.B 5
.IP 3. 5
A common use-case for the numbering by expression is pinning of an application on the Intel Xeon Phi coprocessor with its 60 cores each having 4 SMT threads.
.TP
.B likwid-pin -c E:N:60:1:4 ./myApp
.PP
This command schedules one applicationn thread per physical CPU core for
.B ./myApp.

.SH IMPORTANT NOTICE
The detection of shepard threads works for Intel's/LLVM OpenMP runtime (>=12.0), for GCC's OpenMP runtime as well as for PGI's OpenMP runtime. If you encounter problems with pinning,
please set a proper skip mask to skip the not-detected shepard threads.
Intel OpenMP runtime 11.0/11.1 requires to set a skip mask of
.B 0x1.

.SH AUTHOR
Written by Thomas Gruber <thomas.roehl@googlemail.com>.
.SH BUGS
Report Bugs on <https://github.com/RRZE-HPC/likwid/issues>.
.SH "SEE ALSO"
taskset(1), likwid-perfctr(1), likwid-features(1), likwid-topology(1),


CHANGELOG
# Changelog 5.4.0
- Support for Intel Granite Rapids (core, energy, uncore)
- Support for Intel Sierra Forrest (core, energy, uncore)
- Support for AMD Bergamo (core, energy, uncore)
- Support for Nvidia Grace (core, uncore)
- Fix: AMD Zen4 DataFabric units
- Fix: Multi-socket RAPL measurements on Sapphire Rapids
- Fix: Energy unit for RAPL DRAM domain for SPR, GNR and SRF
- Fix: Discovery mechanism workaround for UPI and M3UPI units of SPR
- Fix: Intel Westmere Uncore with perf_event backend
- Fix: Fujitsu A64FX (more counters, fixed topology, ...)
- Container bridge to use LIKWID inside container
- Sysfeatures interface reworked with support for various architectures and libraries
- Fix build for NVMON for CUDA 12.6+
- likwid-mpirun: SLURM pinning with cpu_mask feature
- Update of internal hwloc (2.11.2) and Lua (5.4.7) version


# Changelog 5.3.0
- Support for Intel SapphireRapids (Core, Uncore, RAPL)
- Support for AMD Zen4 (Core, Uncore, RAPL)
- Support for Apple M1
- Support for AMD GPUs (MarkerAPI, F90 interface)
- Support for AWS Graviton3 (ARM Neoverse V1)
- Support for HiSilicon TSV110
- Fix of F90 interface installation
- Support for extended umasks in ICX and SPR
- Units for metrics in performance groups
- Library calls to get meta information (version, supported features, etc.)
- Some fixes for direct access mode
- Some fixes for X86 RDPMC detection
- Update of internal hwloc (2.9.3) and Lua (5.4.6) version
- New experimental sysfeatures module

# Changelog 5.2.2
- Add mutex to pinning library
- Fix pin string parsing in pinning library
- Make SBIN path configurable in build system
- Add PKGBUILD for ArchLinux package builds
- Remove accessDaemon double-fork in systemd environements
- Group updates for L2/L3 (mainly AMD Zen)
- Fix multi-initialization in MarkerAPI
- Add energy event scaling for Fujitsu A64FX
- Nvmon: Use Cupti error string to get better warning/error messages
- Nvmon: Store events internally to re-use event strings in stopCounters
- AccessLayer: Catch SIGCHLD to stop sending requests to accessDaemon if it was killed
- likwid-genTopoCfg: Update writing and reading of topology file
- Add INST_RETIRED_NOP event for Intel Icelake (desktop & server)
- Removed some memory leaks
- Improved checks for RDPMC availability
- Add TOPDOWN_SLOTS for perf_event
- Fix for systems with CPU sockets without hwthreads (A64FX FX1000)
- Fix if HOME environment variable is not set (systemd)
- Reader function for perf_event_paranoid in Lua to get state early
- likwid-mpirun: Sanitize np and ppn values to avoid crashes


# Changelog 5.2.1
- Add support for Intel Rocketlake and AMD Zen3 variant (Family 19, Model 0x50)
- Fix for perf_event multiplexing (important!)
- Fix for potential deadlock in MarkerAPI (thx @jenny-cheung)
- Build and runtime fixes for Nvidia GPU backend, updates for CUDA test codes
- peakflops kernel for ARMv8
- Updates for AMD Zen1/2/3 event lists and groups
- Support spaces in MarkerAPI region tags (thx @jrmadsen)
- Use 'online' cpulist instead of 'present'
- Switch CI from Travis-CI to NHR@FAU Cx services
- Document -reset and -ureset for likwid-setFrequencies
- Reset cpuset in unpinned runs
- Remove destructor in frequency module
- Check PID if given through --perfpid
- Intel Icelake: OFFCORE_RESPONSE events
- AccessDaemon: Check PCI init state before using it
- likwid-mpirun: Set mpi type for SLURM automatically
- likwid-mpirun: Fix for skip mask for OpenMPI


# Changelog 5.2.0
- Support for AMD Zen3 (Core + Uncore)
- Support for Intel IcelakeSP (Core + Uncore)
- New affinity code
- Fix for Ivybridge uncore code
- Bypass accessdaemon by using rdpmc instruction on x86_64
- Introduce notion of CPU die in topology module
- Use CPU dies for socket-lock for Intel CascadelakeAP
- Add environment variable LIKWID_IGNORE_CPUSET to break out of current CPUset
- Fixes for affinity module CPUlist sorting
- Build against system-installed hwloc
- Update for Intel SkylakeX/CascadelakeX L3 group
- Rename DataFabric events for all generations of AMD Zen
- Add static cache configuration for Fujitsu A64FX
- Add multiplexing checks for perf_event backend
- Fix for table width of likwid-topology afteradding CPU die column
- Adding RasPi 4 with 32 bit OS as ARMv7
- Add default groups for Intel Icelake desktop
- Fix for likwid-setFrequencies to not apply minFreq when setting governor
- likwid-powermeter: Fix hwthread selection when run with -p
- likwid-setFrequencies: Get measured base frequency if register is not readable
- CLOCK group for all AMD Zen
- Fixes in Nvidia GPU support in NvMarkerAPI and topology module

# Changelog 5.1.1
- Support for Intel Cometlake desktop (Core + Uncore)
- Fix for topology module of Fujitsu A64FX
- Fix for Intel Skylake SP in SNC mode
- Fix for likwid-perfscope
- Fix for CLI argument parsing
- Updated group and data file checkers
- Vector sum benchmark in SVE
- FP_PIPE group for Fujitsu A64FX
- Maximal number of CLI arguments configurable in config.mk (currently 16384)

# Changelog 5.1.0
- Support for Intel Icelake desktop (Core + Uncore)
- Support for Intel Icelake server (Core only)
- Support for Intel Tigerlake desktop (Core only)
- Support for Nvidia GPUs with compute capability >= 7.0 (CUpti Profiling API)
- Support for Fujitsu A64FX (Core) including SVE assembly benchmarks
- Support for ARM Neoverse N1 (AWS Graviton 2)
- Support for AMD Zen3 (Core + Uncore but without any events)
- Check for Intel HWP
- Fix for TID filter of Skylake SP LLC filter0 register
- Fix for Lua 5.1
- Fix for likwid-mpirun skip masks
- Fortran90 interface for NvMarkerAPI
- CPU_is_online check to filter non-usable CPU cores

# Changelog 5.0.2
- Fix memory leak in calc_metric()
- New peakflops benchmarks in likwid-bench
- Fix for NUMA domain handling properly
- Improvements for perf_event backend
- Fix for perfctr and powermeter with perf_event backend
- Fix for likwid-mpirun for SLURM with cpusets
- Fix for likwid-setFrequencies in cpusets

src/applications/likwid-pin.lua
#!<INSTALLED_BINPREFIX>/likwid-lua
--[[
 * =======================================================================================
 *
 *      Filename:  likwid-pin.lua
 *
 *      Description:  An application to pin a program including threads
 *
 *      Version:   5.4.0
 *      Released:  15.11.2024
 *
 *      Author:   Thomas Gruber (tr), thomas.roehl@gmail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2024 RRZE, University Erlangen-Nuremberg
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================
]]
package.path = '<INSTALLED_PREFIX>/share/lua/?.lua;' .. package.path

local likwid = require("likwid")

print_stdout = print
print_stderr = function(...) for k,v in pairs({...}) do io.stderr:write(v .. "\n") end end

local function version()
    print_stdout(string.format("likwid-pin -- Version %d.%d.%d (commit: %s)",likwid.version,likwid.release,likwid.minor,likwid.commit))
end

local function examples()
    print_stdout("Examples:")
    print_stdout("There are three possibilities to provide a thread to processor list:")
    print_stdout("1. Thread list with physical thread IDs")
    print_stdout("Example: likwid-pin.lua -c 0,4-6 ./myApp")
    print_stdout("Pins the application to hardware threads 0,4,5 and 6")
    print_stdout("2. Thread list with logical thread numberings in physical cores first sorted list.")
    print_stdout("Example usage thread list: likwid-pin.lua -c N:0,4-6 ./myApp")
    print_stdout("You can pin with the following numberings:")
    print_stdout("\t2. Logical numbering inside node.\n\t   e.g. -c N:0,1,2,3 for the first 4 physical cores of the node")
    print_stdout("\t3. Logical numbering inside socket.\n\t   e.g. -c S0:0-1 for the first 2 physical cores of the socket")
    print_stdout("\t4. Logical numbering inside last level cache group.\n\t   e.g. -c C0:0-3  for the first 4 physical cores in the first LLC")
    print_stdout("\t5. Logical numbering inside NUMA domain.\n\t   e.g. -c M0:0-3 for the first 4 physical cores in the first NUMA domain")
    print_stdout("\tYou can also mix domains separated by  @,\n\te.g. -c S0:0-3@S1:0-3 for the 4 first physical cores on both sockets.")
    print_stdout("3. Expressions based thread list generation with compact processor numbering.")
    print_stdout("Example usage expression: likwid-pin.lua -c E:N:8 ./myApp")
    print_stdout("This will generate a compact list of thread to processor mapping for the node domain")
    print_stdout("with eight threads.")
    print_stdout("The following syntax variants are available:")
    print_stdout("\t1. -c E:<thread domain>:<number of threads>")
    print_stdout("\t2. -c E:<thread domain>:<number of threads>:<chunk size>:<stride>")
    print_stdout("\tFor two hardware threads per core on a SMT4 machine use e.g. -c E:N:122:2:4")
    print_stdout("4. Scatter policy among thread domain type.")
    print_stdout("Example usage scatter: likwid-pin.lua -c M:scatter ./myApp")
    print_stdout("This will generate a thread to processor mapping scattered among all memory domains")
    print_stdout("with physical hardware threads first.")
    print_stdout("")
    print_stdout("likwid-pin sets OMP_NUM_THREADS with as many threads as specified")
    print_stdout("in your pin expression if OMP_NUM_THREADS is not present in your environment.")
end

local function usage()
    version()
    print_stdout("An application to pin a program including threads.\n")
    print_stdout("Options:")
    print_stdout("-h, --help\t\t Help message")
    print_stdout("-v, --version\t\t Version information")
    print_stdout("-V, --verbose <level>\t Verbose output, 0 (only errors), 1 (info), 2 (details), 3 (developer)")
    print_stdout("-i\t\t\t Set numa interleave policy with all involved numa nodes")
    print_stdout("-m\t\t\t Set numa membind policy with all involved numa nodes")
    print_stdout("-S, --sweep\t\t Sweep memory and LLC of involved NUMA nodes")
    print_stdout("-c/-C <list>\t\t Comma separated processor IDs or expression")
    print_stdout("-s, --skip <hex>\t Bitmask with threads to skip")
    print_stdout("-p\t\t\t Print available domains with mapping on physical IDs")
    print_stdout("\t\t\t If used together with -c option outputs the list of physical processor IDs.")
    print_stdout("-d <string>\t\t Delimiter used for using -p to output physical processor list, default is comma.")
    print_stdout("-q, --quiet\t\t Silent without output")
    print_stdout("\n")
    examples()
end

local function close_and_exit(code)
    likwid.putTopology()
    likwid.putAffinityInfo()
    likwid.putConfiguration()
    likwid.unsetenv("LIKWID_NO_ACCESS")
    os.exit(code)
end

delimiter = ','
quiet = 0
sweep_sockets = false
interleaved_policy = false
membind_policy = false
print_domains = false
cpu_list = {}
skip_mask = nil
affinity = nil
num_threads = 0
cpustr = nil
verbose = 0


if (#arg == 0) then
    usage()
    os.exit(0)
end

for opt,arg in likwid.getopt(arg, {"c:", "C:", "d:", "h", "i", "m", "p", "q", "s:", "S", "t:", "v", "V:", "verbose:", "help", "version", "skip","sweep", "quiet"}) do
    if opt == "h" or opt == "help" then
        usage()
        close_and_exit(0)
    elseif opt == "v" or opt == "version" then
        version()
        close_and_exit(0)
    elseif opt == "V" or opt == "verbose" then
        verbose = tonumber(arg)
        likwid.setVerbosity(verbose)
    elseif (opt == "c") or (opt == "C") then
        cpustr = arg
    elseif (opt == "d") then
        delimiter = arg
    elseif opt == "S" or opt == "sweep" then
        sweep_sockets = true
    elseif (opt == "i") then
        interleaved_policy = true
    elseif (opt == "m") then
        membind_policy = true
    elseif (opt == "p") then
        print_domains = true
    elseif opt == "s" or opt == "skip" then
        if arg:match("0x[0-9A-Fa-f]") then
            skip_mask = arg
        else
            if arg:match("[0-9A-Fa-f]") then
                print_stderr("Given skip mask looks like hex, sanitizing arg to 0x"..arg)
                skip_mask = "0x"..arg
            else
                print_stderr("Skip mask must be given in hex")
                close_and_exit(1)
            end
        end
    elseif opt == "q" or opt == "quiet" then
        likwid.setenv("LIKWID_SILENT","true")
        quiet = 1
    elseif opt == "?" then
        print_stderr("Invalid commandline option -"..arg)
        close_and_exit(1)
    elseif opt == "!" then
        print_stderr("Option requires an argument")
        close_and_exit(1)
    end
end
local execList = {}
for i=1, likwid.tablelength(arg)-2 do
    if string.find(arg[i], " ") then
        table.insert(execList, "\""..arg[i].."\"")
    else
        table.insert(execList, arg[i])
    end
end

likwid.setenv("LIKWID_NO_ACCESS", "1")
config = likwid.getConfiguration()
cputopo = likwid.getCpuTopology()
numainfo = likwid.getNumaInfo()
affinity = likwid.getAffinityInfo()

if cpustr ~= nil then
    if not (cpustr:match(",$") or cpustr:match("-$") or cpustr:match("^-") or cpustr:match("^,")) then
        if (affinity ~= nil) then
            num_threads,cpu_list = likwid.cpustr_to_cpulist(cpustr)
        else
            num_threads,cpu_list = likwid.cpustr_to_cpulist_physical(cpustr)
        end
    end
    if (num_threads <= 0) then
        print_stderr("Failed to parse cpulist " .. cpustr)
        close_and_exit(1)
    end
    if verbose > 0 and quiet == 0 then
        print_stdout("Evaluated CPU string to CPUs: " .. cpustr)
    end
end


if print_domains and num_threads > 0 then
    outstr = ""
    for i, cpu in pairs(cpu_list) do
        outstr = outstr .. delimiter .. cpu
    end
    print_stdout(outstr:sub(2,outstr:len()))
    close_and_exit(0)
elseif print_domains then
    for k,v in pairs(affinity["domains"]) do
        print_stdout(string.format("Domain %s:", v["tag"]))
        print_stdout("\t" .. table.concat(v["processorList"], delimiter))
        print_stdout("")
    end
    close_and_exit(0)
end

if num_threads == 0 then
    num_threads, cpu_list = likwid.cpustr_to_cpulist("N")
end
if (#arg == 0) then
    print_stderr("Executable must be given on commandline")
    close_and_exit(1)
end

if interleaved_policy then
    if numainfo["numberOfNodes"] > 1 then
        if verbose > 0 and quiet == 0 then
            print_stdout("Set mem_policy to interleaved")
        end
        likwid.setMemInterleaved(num_threads, cpu_list)
    else
        print_stdout("No need to set mem_policy to interleaved, only one NUMA node available")
    end
end
if membind_policy then
    if numainfo["numberOfNodes"] > 1 then
        if verbose > 0 and quiet == 0 then
            print_stdout("Set mem_policy to membind")
        end
        likwid.setMembind(num_threads, cpu_list)
    else
        print_stdout("No need to set mem_policy to membind, only one NUMA node available")
    end
end

if sweep_sockets then
    if verbose > 0 and quiet == 0 then
        print_stdout("Sweeping memory")
    end
    likwid.memSweep(num_threads, cpu_list)
end

local omp_threads = os.getenv("OMP_NUM_THREADS")
if omp_threads == nil then
    likwid.setenv("OMP_NUM_THREADS",tostring(math.tointeger(num_threads)))
elseif num_threads > tonumber(omp_threads) and (quiet == 0 and verbose > 0) then
    print_stdout(string.format("Environment variable OMP_NUM_THREADS already set to %s but %d cpus required", omp_threads,num_threads))
end
if omp_threads and tonumber(omp_threads) < num_threads then
    num_threads = tonumber(omp_threads)
    for i=#cpu_list,num_threads+1,-1 do
        cpu_list[i] = nil
    end
end

likwid.setenv("KMP_AFFINITY","disabled")
local likwid_hwthreads = {}
for i=1,#cpu_list do
    table.insert(likwid_hwthreads, tostring(math.tointeger(cpu_list[i])))
end
likwid.setenv("LIKWID_HWTHREADS", table.concat(likwid_hwthreads, ","))

if os.getenv("CILK_NWORKERS") == nil then
    likwid.setenv("CILK_NWORKERS", tostring(math.tointeger(num_threads)))
end
if skip_mask then
    likwid.setenv("LIKWID_SKIP", skip_mask)
end

if num_threads > 1 then
    local pinString = tostring(math.tointeger(cpu_list[2]))
    for i=3,likwid.tablelength(cpu_list) do
        pinString = pinString .. "," .. tostring(math.tointeger(cpu_list[i]))
    end
    pinString = pinString .. "," .. tostring(math.tointeger(cpu_list[1]))
    likwid.setenv("LIKWID_PIN", pinString)

    local preload = os.getenv("LD_PRELOAD")
    if preload == nil then
        likwid.setenv("LD_PRELOAD",likwid.pinlibpath)
    else
        likwid.setenv("LD_PRELOAD",likwid.pinlibpath .. ":" .. preload)
    end
    local ldpath = os.getenv("LD_LIBRARY_PATH")
    local libpath = likwid.pinlibpath:match("([^%s]+)/[^%s]+.so")
    if ldpath == nil then
        likwid.setenv("LD_LIBRARY_PATH", libpath)
    elseif libpath and not ldpath:match(libpath) then
        likwid.setenv("LD_LIBRARY_PATH", libpath..":"..ldpath)
    end
else
    likwid.setenv("LIKWID_PIN", tostring(math.tointeger(cpu_list[1])))
    likwid.pinProcess(cpu_list[1], quiet)
end

local exec = table.concat(execList," ")
if verbose > 0 and quiet == 0 then
    print_stdout("Running: " .. exec)
    mask = 0
    for _, c in pairs(cpu_list) do
        -- Check whether the Lua version has bit32 module (>= 5.2)
        if not bit32 then
            mask = mask + 2 ^ c
        else
            mask = bit32.bor(mask, bit32.lshift(1, c))
        end
    end
    print_stdout(string.format("Using %d thread(s) (cpuset: 0x%x)", num_threads, mask))
end
local pid = likwid.startProgram(table.concat(execList," "), num_threads, cpu_list)
if (pid == nil) then
    print_stderr("Failed to execute command: ".. exec)
    close_and_exit(1)
end

local exitvalue = likwid.waitpid(pid)

likwid.putAffinityInfo()
likwid.putTopology()
likwid.putConfiguration()
close_and_exit(exitvalue)
